Add MAX_PATH_ID_BLOCKED frame and related texts (#402)
quicwg/multipath: draft-ietf-quic-multipath.md

as specified in {{Section 9.3 of QUIC-TRANSPORT}} by initiating path validation
but MUST use a new connection ID for the same Path ID.

This proposal adds five multipath control frames for path management:

- PATH_ABANDON frame for the receiver side to abandon the path
(see {{path-abandon-frame}}),
- PATH_STANDBY and PATH_AVAILABLE frames to express a preference
in path usage (see {{path-standby-frame}} and {{path-available-frame}}), and
- MAX_PATH_ID frame (see {{max-paths-frame}}) for increasing the limit of

> active paths, while MAX_PATH_ID_BLOCKED frame (see {{max-paths-blocked-frame}})

Reviewer: ```suggestion
active paths, while PATHS_BLOCKED frame (see {{max-paths-blocked-frame}})
```

> active paths, while PATHS_BLOCKED frame (see {{paths-blocked-frame}})
> indicates that the limit of active paths set by the peer has been reached.

Reviewer: ```suggestion
active paths, and PATHS_BLOCKED frame (see {{paths-blocked-frame}})
to notify the peer of being blocked to open new paths as
the limit of active paths set by the peer has been reached.
```
Is that any better/more accurate?

Reviewer: @gloinul @MikeBishop can you check my proposal above?

> indicates that the limit of active paths set by the peer has been reached.

Reviewer 2: should it be stated that it is an request to the peer to consider increasing it?

Receipt of a frame with a greater Path ID is a connection error as specified
in Section {{frames}}.
When an endpoint finds it has not enough available unused path identifiers,
it SHOULD either send a MAX_PATH_ID frame to increase the active path limit

> (when limited by the sender) or a MAX_PATH_ID_BLOCKED frame

Reviewer: ```suggestion
(when limited by the sender) or a PATHS_BLOCKED frame
```

> (see Section {{max-paths-blocked-frame}}) to inform the peer that a new path

Reviewer: ```suggestion
(see Section {{paths-blocked-frame}}) to inform the peer that a new path
```

A MAX_PATH_ID frame (type=0x15228c0c) informs the peer of the maximum path identifier
it is permitted to use.

When there are not enough unused path identifiers, endpoints SHOULD either send a
MAX_PATH_ID frame to increase the active path limit (when limited by the sender)
or a MAX_PATH_ID_BLOCKED frame to inform the peer that a new path identifier was needed

> but the limit of active paths set by the peer has been reached.

Reviewer: PR #386 removes this sentence here as we should only specify it once in the doc.

a smaller Maximum Path Identifier value than was previously received.
MAX_PATH_ID frames that do not increase the path limit MUST be ignored.

## MAX_PATH_ID_BLOCKED frames {#max-paths-blocked-frame}

A sender SHOULD send a MAX_PATH_ID_BLOCKED frame (type=0x15228c0d) when
it wishes to open a path but is unable to do so due to the maximum path identifier
limit set by its peer;

MAX_PATH_ID_BLOCKED frames are formatted as shown in {{fig-max-paths-blocked-frame-format}}.

~~~
MAX_PATH_ID_BLOCKED Frame {
  Type (i) = 0x15228c0d,
  Maximum Path Identifier (i),
}
~~~
{: #fig-max-paths-blocked-frame-format title="MAX_PATH_ID_BLOCKED Frame Format"}

MAX_PATH_ID frames contain the following field:

Maximum Path Identifier:

> A variable-length integer indicating the maximum number of path identifiers
> allowed at the time the frame was sent. This value MUST NOT exceed 2^32-1
> and MUST NOT be lower than the value advertised in the initial_max_path_id
> transport parameter. Receipt of an invalid Maximum Path Identifier value MUST
> be treated as a connection error of type MP_PROTOCOL_VIOLATION.

Reviewer: ```suggestion
: A variable-length integer indicating the maximum number of path identifiers
  allowed at the time the frame was sent. This value MUST NOT exceed 2^32-1
  and MUST NOT be lower than the value advertised in the initial_max_path_id
  transport parameter. Receipt of an invalid Maximum Path Identifier value MUST
  be treated as a connection error of type MP_PROTOCOL_VIOLATION.
```

> 

Reviewer: This is the same text we already have for the MAX_PATHS frame.

> 

Reviewer 2: I would agree that "invalid" is a bit unclear. Higher than so far declared is fairly obvious. But is closed path ID invalid, they could just be in the process to be closed.

> 

Reviewer 3: My concern is that implementers might consider the PATHS_BLOCKED frame invalid if they have already sent a MAX_PATH_ID with a larger value. Then we get into splitting hair, "lower than initial value is invalid, but lower than current limit is just fine." Might be simpler to say that this frame is ignored with no action if the maximum path identifier parameter is lower than the largest maximum path ID advertised by the receiver.

> 

Reviewer: okay, yes the text was actually not correct as the max limit should be the current max value. a higher value raised an error. However a lower value can just be ignore. Also with the current version of the doc the field carries the max value and not the max number. Please review my proposed update above! @gloinul @huitema @MikeBishop

a smaller Maximum Path Identifier value than was previously received.
MAX_PATH_ID frames that do not increase the path limit MUST be ignored.

> ## MAX_PATH_ID_BLOCKED frames {#max-paths-blocked-frame}

Reviewer: ```suggestion
## PATHS_BLOCKED frames {#paths-blocked-frame}
```

> A sender SHOULD send a MAX_PATH_ID_BLOCKED frame (type=0x15228c0d) when

Reviewer: ```suggestion
A sender SHOULD send a PATHS_BLOCKED frame (type=0x15228c0d) when
```

it wishes to open a path but is unable to do so due to the maximum path identifier
limit set by its peer;

> MAX_PATH_ID_BLOCKED frames are formatted as shown in {{fig-max-paths-blocked-frame-format}}.

Reviewer: ```suggestion
PATHS_BLOCKED frames are formatted as shown in {{fig-max-paths-blocked-frame-format}}.
```

~~~

> MAX_PATH_ID_BLOCKED Frame {

Reviewer: ```suggestion
PATHS_BLOCKED Frame {
```

  Type (i) = 0x15228c0d,
  Maximum Path Identifier (i),
}
~~~
{: #fig-max-paths-blocked-frame-format title="MAX_PATH_ID_BLOCKED Frame Format"}

> MAX_PATH_ID frames contain the following field:

Reviewer: ```suggestion
PATHS_BLOCKED frames contain the following field:
```

Maximum Path Identifier:

> : A variable-length integer indicating the maximum number of path identifiers
>   allowed at the time the frame was sent. This value MUST NOT exceed 2^32-1
>   and MUST NOT be lower than the value advertised in the initial_max_path_id
>   transport parameter. Receipt of an invalid Maximum Path Identifier value MUST
>   be treated as a connection error of type MP_PROTOCOL_VIOLATION.

Reviewer: ```suggestion
: A variable-length integer indicating the maximum path identifiers that was
  allowed at the time the frame was sent. If the received value is lower than
  the currently allowed maximum value, this frame can be ignored.
  Receipt of a value that is larger than the Maximum Path Identifier value MUST
  be treated as a connection error of type PROTOCOL_VIOLATION.
```

TBD-04 (experiments use 0x15228c08)                  | PATH_AVAILABLE      | {{path-available-frame}}
TBD-05 (experiments use 0x15228c09)                  | MP_NEW_CONNECTION_ID   | {{mp-new-conn-id-frame}}
TBD-06 (experiments use 0x15228c0a)                  | MP_RETIRE_CONNECTION_ID| {{mp-retire-conn-id-frame}}
TBD-07 (experiments use 0x15228c0c)                  | MAX_PATH_ID            | {{max-paths-frame}}
TBD-08 (experiments use 0x15228c0d)                  | MAX_PATH_ID_BLOCKED    | {{max-paths-blocked-frame}}

> 

Reviewer: ```suggestion
```

TBD-04 (experiments use 0x15228c08)                  | PATH_AVAILABLE      | {{path-available-frame}}
TBD-05 (experiments use 0x15228c09)                  | MP_NEW_CONNECTION_ID   | {{mp-new-conn-id-frame}}
TBD-06 (experiments use 0x15228c0a)                  | MP_RETIRE_CONNECTION_ID| {{mp-retire-conn-id-frame}}
TBD-07 (experiments use 0x15228c0c)                  | MAX_PATH_ID            | {{max-paths-frame}}

> TBD-08 (experiments use 0x15228c0d)                  | MAX_PATH_ID_BLOCKED    | {{max-paths-blocked-frame}}

Reviewer: ```suggestion
TBD-08 (experiments use 0x15228c0d)                  | PATHS_BLOCKED    | {{max-paths-blocked-frame}}
```
